File: app/src/main/java/org/puffinbasic/runtime/Numbers.java
package org.puffinbasic.runtime;

import static org.puffinbasic.error.PuffinBasicSemanticError.ErrorCode.BAD_NUMBER;

import org.puffinbasic.error.PuffinBasicSemanticError;

import java.util.function.Supplier;

public class Numbers {

    public static int parseInt32(String value, Supplier<String> lineSupplier) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as int32: " + value
            );
        }
    }

    public static int parseInt32(String value, int base, Supplier<String> lineSupplier) {
        try {
            return Integer.parseInt(value, base);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as int32: " + value
            );
        }
    }

    public static long parseInt64(String value, Supplier<String> lineSupplier) {
        try {
            return Long.parseLong(value);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as int64: " + value
            );
        }
    }

    public static long parseInt64(String value, int base, Supplier<String> lineSupplier) {
        try {
            return Long.parseLong(value, base);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as int64: " + value
            );
        }
    }

    public static float parseFloat32(String value, Supplier<String> lineSupplier) {
        try {
            return Float.parseFloat(value);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as float32: " + value
            );
        }
    }

    public static double parseFloat64(String value, Supplier<String> lineSupplier) {
        try {
            return Double.parseDouble(value);
        } catch (NumberFormatException e) {
            throw new PuffinBasicSemanticError(
                    BAD_NUMBER,
                    lineSupplier.get(),
                    "Failed to parse number as float64: " + value
            );
        }
    }
}

File: libbababasic/src/main/java/org/puffinbasic/runtime/Numbers.kt
package org.puffinbasic.runtime

import org.puffinbasic.error.PuffinBasicSemanticError
import java.util.function.Supplier

object Numbers {
    @JvmStatic
    fun parseInt32(value: String, lineSupplier: Supplier<String?>): Int {
        return try {
            value.toInt()
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as int32: $value"
            )
        }
    }

    @JvmStatic
    fun parseInt32(value: String, base: Int, lineSupplier: Supplier<String?>): Int {
        return try {
            value.toInt(base)
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as int32: $value"
            )
        }
    }

    @JvmStatic
    fun parseInt64(value: String, lineSupplier: Supplier<String?>): Long {
        return try {
            value.toLong()
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as int64: $value"
            )
        }
    }

    @JvmStatic
    fun parseInt64(value: String, base: Int, lineSupplier: Supplier<String?>): Long {
        return try {
            value.toLong(base)
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as int64: $value"
            )
        }
    }

    @JvmStatic
    fun parseFloat32(value: String, lineSupplier: Supplier<String?>): Float {
        return try {
            value.toFloat()
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as float32: $value"
            )
        }
    }

    @JvmStatic
    fun parseFloat64(value: String, lineSupplier: Supplier<String?>): Double {
        return try {
            value.toDouble()
        } catch (e: NumberFormatException) {
            throw PuffinBasicSemanticError(
                PuffinBasicSemanticError.ErrorCode.BAD_NUMBER,
                lineSupplier.get()!!,
                "Failed to parse number as float64: $value"
            )
        }
    }
}